### PR TITLE
build: Detect repeated invocations and warn users

### PR DESCRIPTION
**Pull request by evan-charmworks**
Fixes a bug where partial rebuilds in tmp/ lacked -build-shared.
Fixes a bug where re-invoking ./buildold with AMPI-only failed.

### CODE REVIEW

**Comment by evan-charmworks** [comment]
See #1915.

**Review by minitu — APPROVED** [review]

**Comment by matthiasdiener** [comment]
I'm think this feature needs more discussion. I had left it out intentionally from the cmake build for several reasons:
- This usually does not work as expected in normal cmake build systems (you can't re-run cmake in an already configured folder with different options than in the original cmake invocation and expect it to work without removing at least all of cmake's internal state files)
- Instead of re-running the build script or cmake, you can just run `make` in the already configured folder to continue your build (this works even if there were changes to the source code or the build system itself)
- If your `build` command is even slightly different on the second invocation (even adding something innocuous like `--libdir=`), you'll often get an inconsistent build configuration leading to hard-to-debug bugs. We have seen this often in the old build system.

Imho, unless there are strong reasons to have this functionality, it's safer to always remove the build directory.

**Review by matthiasdiener — CHANGES_REQUESTED** [review]
Blocking this until further discussion.

File: buildold [hunk 1]
```diff
@@ -699,14 +704,29 @@ then
...
 # ===================================================
 
 if [ -f "$DESTINATION/tmp/basics" ]
 then
-  Echo "File structure already created; skipping directly to make"
-else
+  Echo "The configure step has already taken place for the specified build."
+  if [ -z "$FORCE" ]
+  then
+    Echo "To attempt a partial rebuild, run make in \"$DESTINATION/tmp\"."
```
Comment (matthiasdiener) [inline buildold:723]:
```suggestion
    Echo "To attempt a partial rebuild, run 'make' in \"$DESTINATION/tmp\"."
```

File: buildcmake [hunk 2]
```diff
@@ -340,6 +341,10 @@ while [[ $# -gt 0 ]]; do
       # Run only cmake, not make
       ONLY_CONFIGURE=1
       ;;
+    --force)
+      # Overwrite existing configuration
```
Comment (matthiasdiener) [inline buildcmake:345]:
```suggestion
      # Overwrite existing build directory
```

File: buildcmake [hunk 3]
```diff
@@ -378,12 +383,55 @@ eval processArgs "$@"
...
+
+
+# Check whether configure step has already taken place
 
+if [[ -f "$builddir/Makefile" ]]
+then
+  echo "The configure step has already taken place for the specified build."
+  if [[ $FORCE -eq 0 ]]
+  then
+    echo "To attempt a partial rebuild, run make in \"$builddir\"."
```
Comment (matthiasdiener) [inline buildcmake:423]:
```suggestion
    echo "To attempt a partial rebuild, run 'make' in \"$builddir\"."
```

**Review by matthiasdiener — APPROVED** [review]

**Review by matthiasdiener — CHANGES_REQUESTED** [review]
Wait - have you tested that this works when pressing Ctrl-C in the middle of the cmake invocation? At that point, the Makefile might not exist, but parts of the configuration have already run and won't be re-executed.

Edit: It might be better to check for presence of `$builddir/CMakeCache.txt`.

**Comment by evan-charmworks** [comment]
> Wait - have you tested that this works when pressing Ctrl-C in the middle of the cmake invocation? At that point, the Makefile might not exist, but parts of the configuration have already run and won't be re-executed.

I've run buildcmake in one terminal with `while true ; do [[ -f 'multicore-linux-x86_64/Makefile' ]] && echo 'Exists' && break ; done` in another and it only came back true once CMake was done executing. Is there any value in a partially-completed configure step? In such a scenario buildcmake will always `rm -rf "$builddir"` and start over.

**Comment by matthiasdiener** [comment]
> > Wait - have you tested that this works when pressing Ctrl-C in the middle of the cmake invocation? At that point, the Makefile might not exist, but parts of the configuration have already run and won't be re-executed.
> 
> I've run buildcmake in one terminal with `while true ; do [[ -f 'multicore-linux-x86_64/Makefile' ]] && echo 'Exists' && break ; done` in another and it only came back true once CMake was done executing. Is there any value in a partially-completed configure step? In such a scenario buildcmake will always `rm -rf "$builddir"` and start over.

Ok, thats a fair point. In case of a partial configure, the config is just silently discarded. My confusion was due to the 'silent' aspect.

**Review by matthiasdiener — APPROVED** [review]

**Review by minitu — APPROVED** [review]

**Comment by evan-charmworks** [comment]
Added an additional check because silently running `rm -rf` sounds like a bad idea, especially with arbitrary `--destination` parameters.

File: buildcmake [hunk 4]
```diff
@@ -378,12 +383,65 @@ eval processArgs "$@"
...
+  echo "The configure step has already taken place for the specified build."
+  if [[ $FORCE -eq 0 ]]
+  then
+    echo "To attempt a partial rebuild, run 'make' in \"$builddir\"."
+    echo "To remove the existing data and reconfigure, run this script again with --force."
+    exit 1
+  else
+    echo "Removing existing data and starting over. (--force)"
+  fi
+elif [[ -e "$builddir" ]]
```
Comment (matthiasdiener) [inline buildcmake:429]: why `-e` and not `-d`? (same below)

Comment (evan-charmworks) [reply buildcmake:429]: If someone mistakenly points --destination at an existing file, for example.

**Review by matthiasdiener — APPROVED** [review]

**Review by stwhite91 — APPROVED** [review]

**Comment by matthiasdiener** [comment]
This PR breaks the spack build:

```bash
$ spack install -j1 -v charmpp@develop
==> Installing charmpp
==> No binary for charmpp found: installing from source
==> charmpp: Executing phase: 'install'
==> [2020-11-14-17:51:05.098606] './build' 'LIBS' 'netlrts-darwin-x86_64' 'clang' 'gfortran' '-j1' '--destination=/Users/mdiener/Work/spack/opt/spack/darwin-macos-skylake/apple-clang-12.0.0/charmpp-develop-5remb5pdx4lkggpjmyd2yxcpnke5khkr' 'smp' '--build-shared' '--with-production'
"/Users/mdiener/Work/spack/opt/spack/darwin-macos-skylake/apple-clang-12.0.0/charmpp-develop-5remb5pdx4lkggpjmyd2yxcpnke5khkr" already exists but is in an inconsistent state.
To remove the existing data and reconfigure, run this script again with --force.
```

**Pull request merged**